Sql database pipeline fixes

steinitzu opened the pull request:
Addressing some of what's mentioned in https://github.com/dlt-hub/pipelines/issues/25  

* Tests for incremental loading and selected tables
* Using `ConnectionStringCredentials`  
* Some restructuring of the code, organized cursor and table loading logic in a separate module

rudolfix commented on tests/sql_database/test_sql_database_pipeline.py:
```diff
@@ -26,14 +30,81 @@ def test_load_sql_schema_loads_all_tables(sql_source_db: SQLAlchemySourceDB, des
+    """Run pipeline twice. Insert more rows after first run
+    and ensure only those rows are stored after the second run.
+    """
+    os.environ['SOURCES__SQL_DATABASE__CHAT_MESSAGE__CURSOR_COLUMN'] = 'updated_at'
```
it is really cool that it works! another thing you can do with dlt that never came to my mind :)

rudolfix commented on pipelines/sql_database/sql_database.py:
```diff
@@ -1,121 +1,43 @@
 
 
 @dlt.source
 def sql_database(
```
the only thing that I feel is missing is the standalone resource. it would call `table_rows` once and let the user to provide the write_disposition, cursor_column and unique_column directly. I gave you a few links on how to do that in previous review

This will be useful for people that want to load a defined set of tables, with different settings and want to declare everything in python, without going to `config.toml` for each table...

rudolfix commented on pipelines/sql_database/sql_database.py:
```diff
@@ -1,121 +1,43 @@
+    credentials: ConnectionStringCredentials = dlt.secrets.value,
     schema: Optional[str] = dlt.config.value,
     table_names: Optional[List[str]] = dlt.config.value,
     write_disposition: str = 'replace',
```
this is still not being used. so in fact all the resources are "append", (or maybe I'm wrong?)
btw. if this is set to replace and the cursor column is specified then the past data will be deleted and only the increment will be kept. there may be use cases for that...

steinitzu replied:
This argument was left over from when I was experimenting initially. I removed it for now. Doesn't really make sense to have it there since you can pass it to `pipeline.run` anyway, right?

rudolfix commented on pipelines/sql_database/util.py:
```diff
@@ -0,0 +1,116 @@
+        if not self.cursor:
+            return query
+        cursor_col = self.cursor_column
+        if cursor_col is None:
```
I'd raise exception if cursor/primary key do not exist

steinitzu replied:
Added this in `Cursor.__init__` so it's raised early as possible

rudolfix reviewed (changes requested):
This is really cool. I still asked for a few improvements. If you want we can merge this version and do the fixes in another PR - or we can continue here

steinitzu commented:
Hey @rudolfix , I added some more updates. Great to get this feedback.  

Added a standalone `sql_table` resource  
And also allowing sqla Engine to be passed instead of credentials. For the table resource that's espcially crucial so you can re-use the connection pool, and when you need something special for the engine setup.  

More tests are needed for the resource and different combinations of cursor/unique column. I'll have to refactor that a bit so there's not so much repetition.   
And will also add batching in `table_rows` result.  
I can do this in another PR.
